packages/hash/api/src/graphql/resolvers/knowledge/entity/before-update-entity-hooks.ts

```diff
@@ -0,0 +1,95 @@
+import { GraphApi } from "@hashintel/hash-graph-client";
+import { ApolloError, UserInputError } from "apollo-server-express";
+import { types } from "@hashintel/hash-shared/types";
+import { WORKSPACE_TYPES } from "../../../../graph/workspace-types";
+import { AccountFields, EntityModel, UserModel } from "../../../../model";
+
+const validateAccountShortname = async (
+  graphApi: GraphApi,
+  shortname: string,
+) => {
+  if (AccountFields.shortnameContainsInvalidCharacter(shortname)) {
+    throw new UserInputError(
+      "Shortname may only contain letters, numbers, - or _",
+    );
+  }
+  if (shortname[0] === "-") {
+    throw new UserInputError("Shortname cannot start with '-'");
+  }
+
+  if (
+    AccountFields.shortnameIsRestricted(shortname) ||
+    (await AccountFields.shortnameIsTaken(graphApi, { shortname }))
+  ) {
+    throw new ApolloError(`Shortname ${shortname} taken`, "NAME_TAKEN");
+  }
+
+  /** @todo: enable admins to have a shortname under 4 characters */
+  if (shortname.length < AccountFields.shortnameMinimumLength) {
+    throw new UserInputError("Shortname must be at least 4 characters long.");
+  }
+  if (shortname.length > AccountFields.shortnameMaximumLength) {
+    throw new UserInputError("Shortname cannot be longer than 24 characters");
+  }
+};
+
+const validateUserPreferredName = (preferredName: string) => {
+  if (UserModel.preferredNameIsInvalid(preferredName)) {
```

Comment:
This would mean the model classes would need to throw GraphQL errors, something we've tried to decouple from the model classes in the past as the model classes aren't always used in environments where GraphQL should be a dependency.

If you feel strongly, we could have the methods throw regular errors, and then catch them in this method to re-throw them as `ApolloError`/`UserInputError` here? But not sure if that's worth the effort.

Reply:
> If you feel strongly, we could have the methods throw regular errors, and then catch them in this method to re-throw them as ApolloError/UserInputError here? But not sure if that's worth the effort.

I think it sounds sensible to do. We need to improve our error handling all over the place it's really not ideal right now. And just saying something like "invalid preferred name" is basically useless without explaining what condition was broken.

Reply:
> just saying something like "invalid preferred name" is basically useless without explaining what condition was broken.

True, but the reason I'm doing this with the preferred name is so that its validation is consistent with the shortname validation. And if you look at `validateUserShortname`, you'll see that the error messages are more descriptive.

> We need to improve our error handling all over the place it's really not ideal right now. 

I agree, but coming up with a consistent approach which works across the various layers seems out of the scope of this PR to me. I'd prefer addressing this in follow-up, and we'd probably benefit from an internal discussion before I do.

Reply:
Yes but as you say `validateAccountShortname` actually does report useful error information, so I don't think this is consistent.

Why don't we change `UserModel.preferredNameIsInvalid` to throw normal errors and then re-throw as the appropriate error here like you said? It doesn't seem like a lot of work

Reply:
> Why don't we change UserModel.preferredNameIsInvalid to throw normal errors and then re-throw as the appropriate error here like you said? It doesn't seem like a lot of work

I think this won't be a simple pattern to get right. We will want to catch all errors, and then re-throw some errors as custom apollo errors with custom codes/messages, and re-throw anything unexpected as `INTERNAL_SERVER_ERROR`s. We should also introduce helper methods to be more consistent with how this works, and if we're already taking the time to refactor how we're throwing things in GraphQL it probably makes sense to get thoughts from FE reviewers. Therefore a standalone PR makes more sense to me, rather than addressing this now.

Reply:
Right but I'm not suggesting we completely rework how we throw ApolloErrors. I'm just suggesting we throw errors from `preferredNameIsInvalid`, that's actually a completely separate concern from GraphQL and Apollo.

I appreciate what you're saying about knowing what to catch though, and in this case it'd just be one thing so I don't really think that's a problem..

If you really don't like the idea of doing it now that's fine, I just don't know if we're really going to catch all of these thoughts in some big task like "rethink errors in GraphQL". It'd be too easy to just take the bool here and throw a different kind of GraphQL error, not even ever thinking about looking at how `UserModel,preferredNameIsInvalid` behaves

Reply:
Have refactored how the preferred name is validated in https://github.com/hashintel/hash/pull/1299/commits/ac23ce59bf0cf644bb90ec8c7d04c474d35fc5d7